### ISSUE TITLE
[Question] Using VecNormalize with expert demonstrations and GAIL (part 3 of 3)

### ISSUE THREAD

**Issue by prabhasak** [hill-a/stable-baselines#931]
Hello. My objective is to imitate an expert using GAIL. I train my custom env on RL, generate experts from the learned model, and train GAIL with this expert. In short: **RL -> expert -> GAIL**. I use [``VecNormalize``](https://stable-baselines.readthedocs.io/en/master/guide/vec_envs.html#vecnormalize) for normalizing DummyEnv observations in all three 'stages' (say DummyEnv_RL, DummyEnv_expert, DummyEnv_IL). That is (following #779):

**RL env:** ``env_RL = VecNormalize(DummyEnv_RL)``
**expert:** ``env_expert = VecNormalize.load(env_RL_path.pkl, env=DummyEnv_expert)``

To train GAIL with the expert, which of the options below is the right thing to do?
**GAIL env:** ``env_GAIL = VecNormalize.load(env_RL_path.pkl, env=DummyEnv_IL)``  **OR**  ``env_GAIL = VecNormalize(DummyEnv_IL)`` 

1. If the answer is ``VecNormalize.load()``, should ``env_GAIL.training`` be true or false (i.e., should ``obs_rms`` stay fixed)?
2. If I were to normalize reward in all the three stages, what should ``env_GAIL.norm.reward`` be?
3. If I train RL and GAIL with different ``gamma``, should I change ``VecNormalize(DummyEnv, gamma)`` accordingly?

**Comment by Miffyli** [comment]
To avoid too much clutter I will close the other related issue and move all discussion here (answers are mostly same).

1) Using GAIL: The implementation here is rather limited, and I highly recommend taking a look at e.g. [this repository on imitation learning](https://github.com/HumanCompatibleAI/imitation). The implementation is in better condition and forced to fit into this stable-baselines scheme.

2) Using different data: There is no plug-n-play to use custom data. You have to take a look at the dataloader for GAIL and create your own for your own data.

3) VecNormalize is mainly required if your observations are arbitrary vectors with unknown basis/distributions, but in your case the MineRL data does not require this (images are normalized automatically already, and other dictionary info needs to be handled other ways).

4) Dictionary obs/actions: Not supported by this version of stable-baselines, but it is planned for stable-baselines3. You need to modify the dict observations/actions into something else (e.g. one-hot encoding).

Other questions are going outside stable-baselines issues (not question related to stable-baselines specifically), and I do not have proper answers for them (e.g. part about gamma).

**Comment by prabhasak** [comment]
> 1\. Using GAIL: The implementation here is rather limited, and I highly recommend taking a look at e.g. [this repository on imitation learning](https://github.com/HumanCompatibleAI/imitation). The implementation is in better condition and forced to fit into this stable-baselines scheme.

Will check this out, thanks :D

> 3\. VecNormalize is mainly required if your observations are arbitrary vectors with unknown basis/distributions, but in your case the MineRL data does not require this (images are normalized automatically already, and other dictionary info needs to be handled other ways).

Ahh, I posted a separate issue _because_ my CustomEnv isn't related to MineRL! (separate project)
Would be great if you could help me with the questions from before :D

> Other questions are going outside stable-baselines issues (not question related to stable-baselines specifically)

Oops, apologies. I rephrased the questions. They're all related to stable-baselines now!

**Comment by Miffyli** [comment]
I think answer for all questions is the same: The normalization parameters should always be the same, because otherwise the algorithms would "see" different observations and/or rewards.  Again, I recommend you check the other repository for examples and code for GAIL specifically.

**Comment by prabhasak** [comment]
Gotcha. The [HumanCompatibleAI](https://github.com/HumanCompatibleAI/imitation) repo seems a little hard to follow (no associated docs yet), but thanks for pointing it out. Closing this!
